scipy/linalg/_matfuncs_expm.pyx.in

```diff
@@ -132,8 +132,8 @@ def pick_pade_structure(cnp.ndarray[ndim=3, dtype=numpy_lapack_t] Am):
     absA = np.abs(Am[0, :, :])
 
     # First spin = normest(|A|, m=1), increase m when spun more
-    np.matmul(absA, work_arr, out=work_arr)
```

Comment:
@thalassemia @ilayn This line I do not understand. We multiply `absA` and `work_arr`, but `work_arr` is initialized to zero on line 116 (or 118) above.

Reply:
Do you mind if I add some changes so we can discuss on it?

Reply:
No problem, go ahead. Although a new branch might be easier because the changes I made in this branch are not valid (I believed `work_arr` to be an n x n array, but it is a n-dimensional vector, so the `absA_pow4` calculation does not make sense at all. there might be a way to re-use `Am[2, :, :]`, but that requires some more thought)

Reply:
`work_arr` is `n` x `n` array. See line 118.

Reply:
Is PyArray_ZERO the one from https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_Zeros? The first argument is `1` on line 118, so the array is 1 dimensional then

Reply:
Ouch... That's a big mistake, probably mine. Can't believe it survived this long. Maybe we got lucky and array is still nxn long but that definitely should be 2.

Reply:
In fact maybe that's why we were getting segfaults all along. This is very helpful @eendebakpt thanks indeed.

Reply:
Should it be 2? I was under the impression we calculate an approximation of the one norm of the matrix powers by repeated multiplication of abs(A) with the 1-dimensional vector [1, 1, 1, 1, ..., 1]. In that way we have multiplications of (n x n) with (n x 1), which is cheaper than a full matrix multiplication. Having the 1 in there means we just take the first element of the `dims` vector which works fine to define the 1d-matrix. 

But this algorithm is new to me, so I'll admit that I do not really know what is happening (or what should be happening)

Reply:
Let me check in the eveningxmaybe i am confusing this completely. I don't remember what I wrote some time ago.

Reply:
Yes you are right, AbsA can't be computed cheaply to the power of 4 and I think you hit the nail on the zero starting array. I will post a PR soon about this. It is also related to #21078. Regardless thank you for improving linalg.

Reply:
@ilayn Looking forward to the PR! Some more ideas:

* For smaller matrices (e.g. n < 10) it might still be worthwhile to compute `absA**4` directly and use that.
* Would it be possible to create a version of `expm` that does not use the `absA` bounds at all? It seems quite a bit faster and accuracy is not reduced a lot (it seems according to the CI that is passing)
* Currently `A**4` is calculated. So we can cheaply compute `abs(A**4)` and with that `abs(A**4)^k * absA**3` as an approximation for `absA**l`. Since `absA**l` was used as a bound for `abs(A)**l` the approximation might work just as good for picking the pade structure. (note: this is just an idea, it would have to proven or checked carefully!)

Reply:
The main block against that is unfortunately |A|**4 can be wildly different than |A**4|. When I was going through the paper to implement this, I also spent quite some time about this ridiculous number of powers but if you think about it, exponentiation of a scalar also quickly blows up similarly. If we don't do this we start chipping away from the accuracy of the result arbitrary amount. 

This time I wrote in C a bit more carefully so hopefully won't take as much time as Cython def function. We are still bounded by the matmuls anyways but at least I managed to shave of some 3 or 4 extra n**2 memory allocations that was in the Cython version, for m=9 and m=13 cases.

Also because we were always having zero as the norm result as you found out, it was always going all the way to the end. Now it will have some easy way out if possible.